make NextCloud WORM file system friendly

kofemann opened the pull request.

kofemann committed:
files: make OC\Files\Storage\Local WORM friendly

Some filesystems run as a Write-Once-Read-Many storages. This
makes them impossible to use with NexeCloud, as the file system
layers uses `truncate` syscall (through file_put_contents function).

As Nextcloud is never updates existing files, removing the old entry
and creatint a new one on update will allow NextCoud to update on such
file systems.

Update Local#fopen and Local#file_put_contents to remote existing
file before truncating.

Signed-off-by: Tigran Mkrtchyan <tigran.mkrtchyan@desy.de>

rullzer commented:
@kofemann as far I know you have been running with this in production for some time now right?
Any reason this PR is still a draft?

kofemann commented:
@rullzer We would like to run some tests to validate that other things (like apps) are not broken. Do you have a testsuite that we can run locally?

PVince81 commented:
the code looks fine to me in general, I'm hoping that the extra unlink will not cause much extra overhead, but we're talking about local storage anyway so should be fine. @icewind1991 thoughts ?

PVince81 commented:
@kofemann please rebase and resolve the conflicts

PVince81 commented:
regarding tests, at least the ones with regular storage, are already running in our CI so if something is broken we'll likely see this immediately

kofemann commented:
@PVince81 Thanks for the update. Actually, since 2 weeks we run this code with nextcloud 22.2.3 in a production. Up to now we haven't seen any issues.

PVince81 commented:
> @PVince81 Thanks for the update. Actually, since 2 weeks we run this code with nextcloud 22.2.3 in a production. Up to now we haven't seen any issues.

that's great to hear! I took the liberty to click "ready for review"

PVince81 commented:
some strange errors in multiple places:
```
1) Test\Cache\FileCacheTest::testGarbageCollectLeaveRecentKeys
InvalidArgumentException: Username is invalid because files already exist for this user
```

I'm wondering if the change in this PR somehow interferes with test cleanup, but I can't see how

kofemann commented:
This is obviously something that I have to look at. Is it possible to run the test manually?

PVince81 commented:
> This is obviously something that I have to look at. Is it possible to run the test manually?

see https://docs.nextcloud.com/server/latest/developer_manual/core/unit-testing.html#bootstrapping-nextcloud

this is how I usually test locally:
`./autotest.sh sqlite tests/lib/Cache/FileCacheTest.php`

icewind1991 commented:
I would expect a filesystem to be smart enough to handle the "truncate an existing file and write to it" case, but I'm fine adding some logic for it.

It should only `unlink` if the file already exists though.

icewind1991 reviewed (changes requested):
Please add a check if the file exists before deleting it

kofemann committed:
files: remove destination file before copying new content (WORM)

Signed-off-by: Tigran Mkrtchyan <tigran.mkrtchyan@desy.de>

kofemann committed:
config: add a switch to control truncate before update

To avoid extra truncate on non WORM file systems, add a new config
option `localstorage.unlink_on_truncate`, which defaults to false.

The OC\Files\Storage\Local is update to respect that option.

Signed-off-by: Tigran Mkrtchyan <tigran.mkrtchyan@desy.de>

kofemann commented:
I have added a coifing option control this behavior. A check before unlink will issue unnecessary stat call.

CarlSchwan commented on lib/private/Files/Storage/Local.php:
```diff
@@ -370,6 +381,9 @@ public function copy($path1, $path2) {
 
 	public function fopen($path, $mode) {
 		$oldMask = umask(022);
+		if ($mode === 'w' && $this->unlinkOnTruncate) {
```
Mode can probably also be 'wb' and 'wb+'. This probably need to be handled too

kofemann replied:
This probably makes sense. However, this is not what we have tested over past several months. I will have a look anyway...

kofemann commented:
rebased on top of current master

PVince81 reviewed (approved):
:+1:

PVince81 commented:
@kofemann the code looks fine. please sign off your commits, see https://github.com/nextcloud/server/pull/24596/checks?check_run_id=8012932504

I'll try and find a second reviewer to get this in. thanks!

CarlSchwan commented on lib/private/Files/Storage/Local.php:
```diff
@@ -69,6 +69,8 @@ class Local extends \OC\Files\Storage\Common {
 
 	private $defUMask;
 
+	protected $unlinkOnTruncate;
```
```suggestion
	protected bool $unlinkOnTruncate;
```

kofemann replied:
This looks reasonable, however other local variables don't explicitly specify the type.

CarlSchwan replied:
Yeah, it is old code, but for new code it's nicer to have typing so that the static analysis can be more powerful

kofemann replied:
👍🏼 fixed
